### PR TITLE
For #12443 - Add firefox suggest header for lib suggestions

### LINKED ISSUE #12443: Firefox Suggest header for library suggestion section of URL bar
**User Story**

Adding the 'Firefox Suggest' branded term to URL bar suggestions pulled from library data is intended to:

Emphasize to users that library data suggestions are being generated by Firefox, and are distinct from default search suggestions

Lay the groundwork for Firefox Suggest, which is intended to further increase the value users derive from Firefox suggested results for partial query strings

Link to [spec](https://docs.google.com/document/d/1HkD1WkpbGhH1mVTtn16Ako8-tCWV1XSEY58rvUa8jYY/edit?usp=sharing), and design [mockup](https://www.figma.com/file/epu0Sj6mCPk7K282sNpHgI/Site?node-id=3248%3A50053) for reference.

Please disregard the “Sponsored” components of the design mockup, including sponsored suggestions and related settings options. Those items will be addressed in a separate issue.

**Acceptance Criteria**

The branded term of “Firefox Suggest” is included as a header for the section of Fenix URL bar suggestions pulled from user library data (i.e. suggested bookmarks, history, open tabs).

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-20931)

### PR DESCRIPTION
**Pull request by sarah541**
Fixes #12443 
This is a generalized way to group suggestions based on suggestion providers. In future, if needed, suggestions can be grouped by just providing the same header name.

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

Fixes #12443
Fixes #12443
Fixes #12443
Fixes #12443
Fixes #12443

### CODE REVIEW

File: components/ui/icons/src/main/res/values/colors.xml [hunk 1]
```diff
@@ -1,4 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
 <resources>
     <color name="mozac_ui_icons_fill">#FFFFFF</color>
+    <color name="photonDarkGrey05">#ff5b5b66</color>
```
Comment (gabrielluong) [inline components/ui/icons/src/main/res/values/colors.xml:4]: This doesn't look like it is used.

File: components/feature/awesomebar/src/main/java/mozilla/components/feature/awesomebar/provider/SearchSuggestionProvider.kt [hunk 2]
```diff
@@ -37,7 +37,7 @@ class SearchSuggestionProvider private constructor(
     @VisibleForTesting internal val engine: Engine? = null,
     private val icon: Bitmap? = null,
     private val showDescription: Boolean = true,
-    private val filterExactMatch: Boolean = false
+    private val filterExactMatch: Boolean = false,
```
Comment (gabrielluong) [inline components/feature/awesomebar/src/main/java/mozilla/components/feature/awesomebar/provider/SearchSuggestionProvider.kt:40]: I would avoid all the formatting changes here since they aren't needed for your issue. It's better to preserve the existing git history.

File: components/concept/awesomebar/src/main/java/mozilla/components/concept/awesomebar/AwesomeBar.kt [hunk 3]
```diff
@@ -145,6 +145,10 @@ interface AwesomeBar {
      * It returns a list of [Suggestion]s to be displayed by the [AwesomeBar].
      */
     interface SuggestionProvider {
+        /**
+         * A header title for grouping the suggestions.
+         **/
+        fun groupTitle(): String? = ""
```
Comment (gabrielluong) [inline components/concept/awesomebar/src/main/java/mozilla/components/concept/awesomebar/AwesomeBar.kt:151]: Can you add a new line before this? It might be better to move this below `id`. It feels like ID is usually one of those starting parameters you would see.

File: components/compose/awesomebar/src/main/java/mozilla/components/compose/browser/awesomebar/internal/SuggestionGroup.kt [hunk 4]
```diff
@@ -23,6 +25,11 @@ internal fun SuggestionGroup(
         title,
         color = colors.groupTitle,
         modifier = Modifier
-            .padding(4.dp)
+            .padding(
```
Comment (rocketsroger) [inline components/compose/awesomebar/src/main/java/mozilla/components/compose/browser/awesomebar/internal/SuggestionGroup.kt:28]: Is it possible for this to be dynamic? In case the application side did not pass in a group title?

File: components/feature/awesomebar/src/main/java/mozilla/components/feature/awesomebar/provider/SearchEngineSuggestionProvider.kt [hunk 5]
```diff
@@ -24,6 +24,7 @@ import java.util.UUID
  * @property searchIcon the icon to be shown for the suggestion(s), same icon for all
  * @property maxSuggestions the maximum number of suggestions to be provided
  * @property charactersThreshold the minimum typed characters used to match to a search engine name
+ * @property suggestionsHeader optional parameter to specify if the suggestion should have a header
```
Comment (rocketsroger) [inline components/feature/awesomebar/src/main/java/mozilla/components/feature/awesomebar/provider/SearchEngineSuggestionProvider.kt:27]: This should have been there.  But since you're not touching anything in this file, can we please do this in a separate PR?  Thanks

File: components/feature/awesomebar/src/main/java/mozilla/components/feature/awesomebar/provider/SearchActionProvider.kt [hunk 6]
```diff
@@ -25,7 +26,7 @@ class SearchActionProvider(
     private val showDescription: Boolean = true,
     private val searchEngine: SearchEngine? = null
 ) : AwesomeBar.SuggestionProvider {
-    override val id: String = java.util.UUID.randomUUID().toString()
+    override val id: String = UUID.randomUUID().toString()
```
Comment (rocketsroger) [inline components/feature/awesomebar/src/main/java/mozilla/components/feature/awesomebar/provider/SearchActionProvider.kt:29]: Same with this one, I like the change but maybe in a separate PR with the one below?

File: components/concept/awesomebar/src/main/java/mozilla/components/concept/awesomebar/AwesomeBar.kt [hunk 7]
```diff
@@ -152,6 +152,11 @@ interface AwesomeBar {
          */
         val id: String
 
+        /**
+         * A header title for grouping the suggestions.
+         **/
+        fun groupTitle(): String? = ""
```
Comment (rocketsroger) [inline components/concept/awesomebar/src/main/java/mozilla/components/concept/awesomebar/AwesomeBar.kt:158]: Should the default value here be null? Since the the return is `String?`

File: components/feature/awesomebar/src/main/java/mozilla/components/feature/awesomebar/provider/BookmarksStorageSuggestionProvider.kt [hunk 8]
```diff
@@ -39,10 +40,14 @@ class BookmarksStorageSuggestionProvider(
     private val indicatorIcon: Drawable? = null,
     private val engine: Engine? = null,
     private val showEditSuggestion: Boolean = true,
+    private val suggestionsHeader: String? = "",
```
Comment (rocketsroger) [inline components/feature/awesomebar/src/main/java/mozilla/components/feature/awesomebar/provider/BookmarksStorageSuggestionProvider.kt:43]: Should the default here be null? (Same with all similar cases below)

**Review by rocketsroger — COMMENTED** [review]
Looks good, almost there.  I have some questions.  And I think couple files can be removed from this review.  Thanks

File: components/compose/awesomebar/src/main/java/mozilla/components/compose/browser/awesomebar/internal/SuggestionGroup.kt (hunk 4 above)
Comment (sarah541) [reply components/compose/awesomebar/src/main/java/mozilla/components/compose/browser/awesomebar/internal/SuggestionGroup.kt:28]: SuggestionGroup composable is never called if the group title is empty. See below

`group.title?.let { title ->
                 tem(group.id) {
                        SuggestionGroup(title, colors)
`

**Review by rocketsroger — APPROVED** [review]
Looks good! 🚢

**Comment by Mugurell** [comment]
@sarah541 Tested with before and after builds and saw that after this patch the suggestions are flickering:


https://user-images.githubusercontent.com/11428869/179259027-fffa1f51-8a3a-431d-a3db-8bcefad0ff4f.mp4

File: components/compose/awesomebar/src/main/java/mozilla/components/compose/browser/awesomebar/AwesomeBar.kt [hunk 9]
```diff
@@ -41,10 +40,22 @@ fun AwesomeBar(
...
+    }
+    val groups: ArrayList<AwesomeBar.SuggestionProviderGroup> = ArrayList()
+    for (item in mapTitlesToProviders) {
+        groups.add(
             AwesomeBar.SuggestionProviderGroup(
-                providers = providers
+                providers = item.value,
+                title = item.key
             )
         )
```
Comment (Mugurell) [inline components/compose/awesomebar/src/main/java/mozilla/components/compose/browser/awesomebar/AwesomeBar.kt:60]: If the providers and the group titles don't change we could keep the old `remember` functionality to avoid recompositions with something like:

```suggestion
val groups = remember(providers) {
    providers
        .groupBy { it.groupTitle() }
        .map {
            AwesomeBar.SuggestionProviderGroup(
                providers = it.value,
                title = it.key
            )
        }
}
```

Comment (sarah541) [reply components/compose/awesomebar/src/main/java/mozilla/components/compose/browser/awesomebar/AwesomeBar.kt:60]: Thank you @Mugurell !
